runtime/major_gc.c

```diff
@@ -1053,6 +1052,7 @@ static void cycle_all_domains_callback(struct domain* domain, void* unused,
   domain->state->final_info->updated_first = 0;
   domain->state->final_info->updated_last = 0;
 
+  caml_global_barrier();
```

Comment:
Might be nice to signpost that we need this for correctness in a comment, at the very least for the global root marking.